mls-rs-crypto-webcrypto/src/ec/der_private_key.rs

```diff
@@ -0,0 +1,135 @@
+    84, 94, 56, 114, 118, 10, 183, 54, 23, 222, 74, 150, 38, 44, 111, 93, 158, 152, 191, 146, 146,
+    220, 41, 248, 244, 29, 189, 40, 154, 20, 124, 233, 218, 49, 19, 181, 240, 184, 192, 10, 96,
+    177, 206, 29, 126, 129, 157, 122, 67, 29, 124, 144, 234, 14, 95,
+];
+
+const P_512_GENERATOR: &[u8] = &[
+    4, 0, 198, 133, 142, 6, 183, 4, 4, 233, 205, 158, 62, 203, 102, 35, 149, 180, 66, 156, 100,
+    129, 57, 5, 63, 181, 33, 248, 40, 175, 96, 107, 77, 61, 186, 161, 75, 94, 119, 239, 231, 89,
+    40, 254, 29, 193, 39, 162, 255, 168, 222, 51, 72, 179, 193, 133, 106, 66, 155, 249, 126, 126,
+    49, 194, 229, 189, 102, 1, 24, 57, 41, 106, 120, 154, 59, 192, 4, 92, 138, 95, 180, 44, 125,
+    27, 217, 152, 245, 68, 73, 87, 155, 68, 104, 23, 175, 189, 23, 39, 62, 102, 44, 151, 238, 114,
+    153, 94, 244, 38, 64, 197, 80, 185, 1, 63, 173, 7, 97, 53, 60, 112, 134, 162, 114, 194, 64,
+    136, 190, 148, 118, 159, 209, 102, 80,
+];
+
+// Generate private / public key pair.
+pub(crate) async fn generate(curve: &'static str) -> Result<(Vec<u8>, Vec<u8>), CryptoError> {
+    let crypto = get_crypto()?;
+    let private_key_type = KeyType::EcdhSecret(curve);
+    let public_key_type = KeyType::EcdhPublic(curve);
+
+    let params = EcKeyGenParams::new(private_key_type.algorithm(), curve);
+
+    let key_usages = Array::new_with_length(1);
+    key_usages.set(0, private_key_type.usage().into());
+
+    let key_pair = crypto.generate_key_with_object(&params, true, &key_usages)?;
+    let key_pair: CryptoKeyPair = JsFuture::from(key_pair).await?.into();
+
+    let private_key = private_key_type.export(&crypto, &key_pair).await?;
+    let public_key = public_key_type.export(&crypto, &key_pair).await?;
+
+    Ok((private_key, public_key))
+}
+
+#[derive(Debug, Sequence)]
+pub(crate) struct DerPrivateKey {
+    version: Uint,
+    pub private_key: OctetString,
+    pub public_key: Option<ContextSpecific<BitString>>,
+}
+
+// `private_key` is a DER-serialized `DerPrivateKey` struct.
+#[derive(Debug, Sequence)]
+struct DerPrivateKeySerializedFormat {
+    version: Uint,
+    algorithm_identifier: AlgorithmIdentifier,
+    private_key: OctetString,
+}
+
+#[derive(Debug, Sequence, Clone)]
+struct AlgorithmIdentifier {
+    object_identifier: ObjectIdentifier,
+    parameters: Any,
+}
+
+impl DerPrivateKey {
+    // Generates a key without the public key.
+    pub(crate) fn from_raw(private_key: &[u8]) -> Result<Self, der::Error> {
+        Ok(Self {
+            version: Uint::new(&[1])?,
+            private_key: OctetString::new(private_key)?,
+            public_key: None,
+        })
+    }
+
+    // Replaces the public key by the group generator. Can be used when an implementation
+    // requires some public key but the public key does not matter for the operation at
+    // hand, such as ECDH and ECDSA with NIST curves.
+    pub(crate) fn set_fake_public_key(&mut self, curve: Curve) {
```

Comment:
This is required for Chrome to work, correct?

Reply:
It's not. It now allows Firefox to do HPKE decryption assuming it's given its own public key, e.g. taken unchecked from the update path. It'll also be needed when we support Firefox once they support  compressed public key format. I thought adding this doesn't hurt, but I can remove it for now?